Fix #1631: Add audit user documentation (#1636)
wultra/enrollment-server: docs/onboarding/Audit.md

# Audit

This feature provides audit logging for the onboarding server in the database.

> For more detailed developer documentation see [auditing library documentation](https://github.com/wultra/java-core?tab=readme-ov-file#wultra-auditing-library).

Author: We already have developer documentation at this link. The purpose of this documentation is for users who are not necessarily developers.
At this moment, it is written for the onboarding server, but we can easily transform it into a general auditing guide, since the same library is used across components.
Only the **Audit types** section should be specific to the onboarding server

| `message`           | The message of the audit log entry                                                          | -                |
| `exception_message` | Error message if specified for the log entry                                                | -                |
| `stack_trace`       | Error stack trace if specified for the log entry                                            | -                |
| `param`             | Extra parameters for the log entry (independent of the `message`)                           | -                |
| `calling_class`     | The calling Java class of the audit log entry                                               | 256              |
| `thread_name`       | The thread producing the log entry                                                          | 256              |
| `version`           | Version of the application                                                                  | 256              |
| `build_time`        | The build time of the application                                                           | -                |

### Table `audit_param`

> Table for audit log parameters. It contains the same information as `audit_log.param`, split by parameter name into records into this table.

Author: Fixed

Storing data in this table is disabled by default and is controlled by the `audit.db.table.param.enabled` property.

<!-- begin box warning -->
The column `audit_log.param` does not have a length limit, but the column `audit_param.param_value` is limited to 4,000 characters. 
If a value exceeding this limit is passed, it is truncated to 4,000 characters. The same applies to the `audit_param.param_key` where the limit is 256 characters.
<!-- end -->

| Column               | Description                                                         | Characters limit |
|----------------------|---------------------------------------------------------------------|------------------|
| `audit_log_id`       | The identifier of the audit log entry from `audit_log.audit_log_id` | 36               |
| `timestamp_created`  | Time when the parameter record was created                          | -                |

> | `param_name`         | The name of the parameter from `audit_log.param`.                   | 256              |

Author: Fixed

| `param_value`       | The value of the parameter from `audit_log.param`.                  | 4000             |

> ## Configuration

Author: At the user documentation level, we may not need to reveal and describe all the configuration parameters. I’ve mentioned all of them, and we can reduce them at any time.

| `identityVerification`         | Identity verification operations                                                                                                                                          |
| `activation`                   | Activation operations                                                                                                                                                     |
| `documentVerification`         | Document verification operations                                                                                                                                          |
| `presenceCheckProvider`        | Presence check operations                                                                                                                                                 |
| `documentVerificationProvider` | Document verification provider operations                                                                                                                                 |

> | `onboardingProvider`           | [Onboarding provider](https://developers.wultra.com/components/enrollment-server/develop/documentation/onboarding/External-Onboarding-Services#api-endpoints) operations  |

Reviewer: Shouldn't we be able to use a relative URL here?

Author: Yes and the relative url is definitely better, fixed

## Audit levels

A value in the `audit_log.audit_level` column is used to categorize the audit log entry according to severity.

Values from the table below are supported. The table is sorted from the lowest logging level to the highest. Selecting a log level 
includes all levels that precede it. For example, if the level `INFO` is configured, all messages with levels `ERROR`, `WARN` and `INFO` are persisted, 
while messages with levels `DEBUG` and `TRACE` are ignored and not persisted.

| Value    | Description                                               |
|----------|-----------------------------------------------------------|

> | `NONE`   | No log message ire persisted - audit logging is disabled. |

Author: Fixed

| `ERROR`  | Error level                                              |
| `WARN`   | Warning level                                            |
| `INFO`   | Information level                                        |
| `DEBUG`  | Debug level                                              |
| `TRACE`  | Trace level                                              |
| `ALL`    | All levels - same effect as `TRACE`                      |

> ## How to search in audit logs

Reviewer: I suggest adding a warning box about performance and explaining that the audit entries are supposed to be used for investigation only, not for reporting, as someone tends to do so.

Author: Good idea, I added the warning and an example of the query result in CSV format

| `audit.event.queue.size`                         | `100000`             | Size of the internal queue for log entries to be written.                                           |
| `audit.db.cleanup.days`                          | `365`                | Data retention period in the database.                                                              |
| `audit.db.table.log.name`                        | `audit_log`          | Name of the database table for audit log entries.                                                   |
| `audit.db.table.param.enabled`                   | `false`              | Whether values from the column `audit_log.param` are parsed and persisted into `audit_param` table. |
| `audit.db.table.param.name`                      | `audit_param`        | Name of the database table for audit log parameters.                                                |
| `spring.jpa.properties.hibernate.default_schema` | _empty_              | The database schema to use.                                                                         |

> | `audit.db.batch.size`                            | `1000`               | Batch size for database operations.                                                                 |

Reviewer 2: - the database structure is already described in database structure (with link to the Audit core). So basically, remove redundancies up to line 67

Author: The section was removed

## Audit types

A value in the `audit_log.audit_type` column is used to categorize the audit log entry according to the operation scope.
The following values are used:

| Value                          | Description                                           |
|--------------------------------|-------------------------------------------------------|
| `process`                      | General onboarding process operation                  |
| `otp`                          | OTP operations                                        |
| `identityVerification`         | Identity verification operations                      |
| `activation`                   | Activation operations                                 |
| `documentVerification`         | Document verification operations                      |
| `presenceCheckProvider`        | Presence check operations                             |
| `documentVerificationProvider` | Document verification provider operations             |

> | `onboardingProvider`           | [Onboarding provider](./Onboarding-API.md) operations |

Reviewer 3: I don't understand why there is a link to the API when it is only used in asynchronous evaluation/approval?

Author: Sorry, you are right the mentioned API is not correct, there should be the one for [External Onboarding Services](https://developers.wultra.com/components/enrollment-server/develop/documentation/onboarding/External-Onboarding-Services#api-endpoints). Fixed

> ## Audit levels

Reviewer 2: Unless we are using the levels in audit, remove the chapter. Or at least the unused levels.

Author: The section was removed

| `ERROR`  | Error level                                              |
| `WARN`   | Warning level                                            |
| `INFO`   | Information level                                        |
| `DEBUG`  | Debug level                                              |
| `TRACE`  | Trace level                                              |
| `ALL`    | All levels - same effect as `TRACE`                      |

## How to search in audit logs

<!-- begin box warning -->
The audit log should be used to investigate specific onboarding processes. It should not be used for reporting, as it is not optimized for this purpose 
and would negatively impact performance.
<!-- end -->

The most frequently logged parameters are `activationId` and `userId`. Values are stored either in the `message` or `param` column of the `audit_log` table.

For example, to find all entries with activation ID `6c115802-fad3-474a-afe6-b69215740a99` following query can be used:

```sql
SELECT *
FROM audit_log
WHERE message LIKE '%6c115802-fad3-474a-afe6-b69215740a99%'
OR param LIKE '%6c115802-fad3-474a-afe6-b69215740a99%'
```

> Example of a result for the query above in CSV format:

Reviewer 2: Example is fine but there is no documentation what are we actually auditing in terms of parameters...

Author: Added description for each event

# Audit

This feature provides audit logging for the onboarding server in the database.
For more detailed developer documentation see [auditing library documentation](https://github.com/wultra/java-core?tab=readme-ov-file#wultra-auditing-library).

<!-- begin box warning -->
The audit log should be used to investigate specific onboarding processes. It should not be used for reporting, as it is not optimized for this purpose
and would negatively impact performance.
<!-- end -->

A value in the `audit_log.audit_type` column is used to categorize the audit log entry according to the operation scope.
The following values are used:

> | Value                          | Description                                                                                                                                                                                                                            |

Author: I am thinking this table may not be necessary, and each type could be described directly in the `## Event type {type}` section. What is your opinion?

Reviewer 2: Yes,
You are right. 
Like, there is certainly a point to have a list of events in one place, and the deeper description elsewhere. 

This way a reader can see all types that are logged - to grasp the basic patterns of audit, and read the details of what the events actually mean - this is required.

Author: I removed the table—the description for each type was moved into the chapter, and I left only the list here.

|------------------------------|---------------------------------------------------------------------------------------------------|
| `processId`                  | ID of the onboarding process                                                                      |
| `activationId`               | PowerAuth activation ID bind with the onboarding process                                          |
| `userId`                     | ID of user passing the onboarding process                                                         |
| `documentVerificationStatus` | Status of document verification. Possible values: `IN_PROGRESS`, `ACCEPTED`, `REJECTED`, `FAILED` |

## Event type `process`

### Message `Process started for user: {userId}`

The onboarding process was created and started. The endpoint `POST api/onboarding/start` was called and existing process was not found for the provided user identification data.

### Message `Process resumed for user: {userId}`

The onboarding process was resumed. The endpoint `POST api/onboarding/start` was called and existing process was found for the provided user identification data.

### Message `Process finished for user: {userId}`

> The onboarding process was successfully finished.

Reviewer 2: I would be even more specific. Does this mean the process was set to the same status and phase?
Taking a data-point perspective (i.e., status and phase) can be more beneficial than an API perspective.
The API point of view is good and can be retained; just the data point of view is better.

Author: Added phase and status, explaining that this is a terminal state and is expected to occur at most once for the process.

### Message `Process failed: {errorDetail}, for user: {userId}`

> The onboarding process failed. See detail in the `errorDetail` parameter.

Reviewer 2: Same as above.
BTW, from an audit point of view, it's important to understand whether I can have multiple audit records for one process.

Author: Same as above

### Message `Expired process for user: {userId}, {errorDetail}`

> The onboarding process was not completed in a given time, and the process is cleaned up by a scheduled job.

Reviewer 2: ```suggestion
The onboarding process is not completed within a given time, and the process is cleaned up by a scheduled job.
```

Or use past tense in both sentences:
The onboarding process was not completed within a given time, and the process was cleaned up by a scheduled job.

Reviewer 2: I see we are using the past tense so stick to past tense.

Author: Fixed by using the past tense

### Message `OTP failed for user: {userId}`

Previously verified user verification OTP was invalidated and marked as failed. This happens when the OTP step succeeds,
but the combined SCA evaluation fails because the earlier presence check result was not successful.

### Message `Expired OTP for user: {userId}`

The OTP was not used in a given time, and the process is cleaned up by a scheduled job.

### Message `OTP expired for user: {userId}`

An OTP code was submitted after its expiration time and could no longer be verified.

### Message `OTP {otpType} verified for user: {userId}`

The user submitted valid OTP code.

### Message `OTP {otpType} verification failed for user: {userId}`

The user submitted invalid OTP code.

### Message `OTP max attempts reached for user: {userId}`

Too many OTP verification attempts were made.

### Message `OTP failed because of failed process for user: {userId}`

The onboarding process is failed and the OTP verification is not possible.

### Message `Resending OTP for user: {userId}`

The user requested to resend the OTP code.

### Message `OTP canceled for process ID: {processId}, user ID: {userId}, otp type: {otpType}`

> Invalidate the OTP code. This is done by calling the endpoint `POST /api/onboarding/cleanup`.

Reviewer 2: ```suggestion
The OTP code is invalidated by calling the endpoint `POST /api/onboarding/cleanup`.
```

Author: Fixed

### Message `Cross verified documents: {documentVerificationStatus} for user: {userId}`

Result of the final verification for all verified documents.

### Message `Documents verified: {documentVerificationStatus} for user: {userId}`

The provider returned evaluation for an entire set of uploaded documents. The set means all documents in the request body of
the upload endpoint `POST /api/v2/identity/document/submit`.

### Message `Check document upload for user: {userId}`

A face photo of a person was fetched from a trusted source.

### Message `Sdk initialized for user: {userId}`

The mobile client document verification SDK is initialized and ready to be used for scanning identity documents.
The endpoint `POST /api/identity/document/init-sdk` was called.

### Message `Cleaned up documents for user: {userId}`

Documents were deleted from the document verification provider and server during manual cleanup by calling the `POST /api/identity/cleanup` endpoint.

## Event type `onboardingProvider`

### Message `Client evaluated for user: {userId}`

> Client evaluation was successfully sent to the [Client evaluation service](./External-Onboarding-Services.md#client-evaluation-service).

Reviewer 2: ```suggestion
The client evaluation was successfully sent to the [Client evaluation service](./External-Onboarding-Services.md#client-evaluation-service).
```

Author: Fixed

### Message `Sent user verification OTP for user: {userId}`

> User verification OTP was sent to the onboarding provider for the first time.

Reviewer 2: ```suggestion
A new user verification OTP was sent to the onboarding provider.
```

Author: Fixed

### Message `Resent user verification OTP for user: {userId}`

> User verification OTP was resent to the onboarding provider.

Reviewer 2: ```suggestion
The user verification OTP was resent to the onboarding provider.
```

Author: Fixed

# Audit

This feature provides audit logging for the onboarding server in the database.
For more detailed developer documentation see [auditing library documentation](https://github.com/wultra/java-core?tab=readme-ov-file#wultra-auditing-library).

## Configuration

> The following application properties can be used to configure audit logging:

Reviewer 2: delete section

Author: Deleted

| `audit.db.table.log.name`                        | `audit_log`          | Name of the database table for audit log entries.                                                   |
| `audit.db.table.param.enabled`                   | `false`              | Whether values from the column `audit_log.param` are parsed and persisted into `audit_param` table. |
| `audit.db.table.param.name`                      | `audit_param`        | Name of the database table for audit log parameters.                                                |
| `spring.jpa.properties.hibernate.default_schema` | _empty_              | The database schema to use.                                                                         |
| `audit.db.batch.size`                            | `1000`               | Batch size for database operations.                                                                 |

## Audit types

A value in the `audit_log.audit_type` column is used to categorize the audit log entry according to the operation scope.
The following values are used:

> | Value                          | Description                                                         |

Reviewer 2: Too minimalistic descriptions. Please expand.

Author: Explained scope and operations for each type

| `otp`                          | OTP operations                                                      |
| `identityVerification`         | Identity verification operations                                    |
| `activation`                   | Activation operations                                               |
| `documentVerification`         | Document verification operations                                    |
| `presenceCheckProvider`        | Presence check operations                                           |
| `documentVerificationProvider` | Document verification provider operations                           |
| `onboardingProvider`           | [Onboarding provider](./External-Onboarding-Services.md) operations |

## How to search in audit logs

<!-- begin box warning -->
The audit log should be used to investigate specific onboarding processes. It should not be used for reporting, as it is not optimized for this purpose 
and would negatively impact performance.
<!-- end -->

The most frequently logged parameters are `activationId` and `userId`. Values are stored either in the `message` or `param` column of the `audit_log` table.

> For example, to find all entries with activation ID `6c115802-fad3-474a-afe6-b69215740a99` following query can be used:

Reviewer 2: This is way to crude.  And we will change this before release so feel free to delete this.

Author: That’s great we will take a different approach - the query has been removed.

```sql
SELECT *
FROM audit_log
WHERE message LIKE '%6c115802-fad3-474a-afe6-b69215740a99%'
   OR param LIKE '%6c115802-fad3-474a-afe6-b69215740a99%'
order by timestamp_created desc;
```

Example of the query result for a successful onboarding process. Events are sorted chronologically—the newest first:

```csv

> audit_log_id,application_name,audit_level,audit_type,timestamp_created,message,exception_message,stack_trace,param,calling_class,thread_name,version,build_time

Reviewer 2: Examples are fine, but any example has to be shown AFTER the structure is explained.

Author: The example is now very last section

8da113bd-dbeb-4c35-9e41-4d88e10e7628,onboarding-server,INFO,identityVerification,2026-03-19 11:53:13.818000,Switched to DOCUMENT_UPLOAD/IN_PROGRESS; user ID: mockuser_567780649667363606,,,"{""identityVerificationId"":""189db1e4-7abe-4309-a983-19e6283e3496"",""processId"":""8d0659cb-9e8c-401a-9493-2284f52c5528"",""activationId"":""6c115802-fad3-474a-afe6-b69215740a99"",""userId"":""mockuser_567780649667363606""}",com.wultra.app.onboardingserver.common.service.AuditService,http-nio-8083-exec-3,2.1.0-SNAPSHOT,2026-03-19 11:51:59.726000
```

> ### Parameters

Reviewer 2: We can have section common parameters (show for all event types)
And then section what is shown for a specific  event_type.

In your case I would use this breakdown:

# Common parameters

# Event type `process`
Table of specific parameters for the event type `process` (exluding the common)
## Message 1
Table of specific parameters for this message 1 (excluding the common and event type specific)
## Message 2
Table of specific parameters for this message 2 (excluding the common and event type specific)

Author: I used the structure

# Audit

This feature provides audit logging for the onboarding server in the database.
For more detailed developer documentation see [auditing library documentation](https://github.com/wultra/java-core?tab=readme-ov-file#wultra-auditing-library).

## Database structure

The auditing library uses the following database structure. 

> 

Author: Remove the `Database structure` section

If a value longer than the character limit is passed to a column, it is truncated to the limit.

## Configuration

The following application properties can be used to configure audit logging:

| Property Name                                    | Default Value        | Description                                                                                         |
|--------------------------------------------------|----------------------|-----------------------------------------------------------------------------------------------------|
| `spring.application.name`                        | `onboarding-server`  | The name of application used to set `application_name` column                                       |
| `audit.level`                                    | `INFO`               | Threshold of log message levels to be persisted. See the [Audit levels](#audit-levels) section.     |

> | `audit.event.queue.size`                         | `100000`             | Size of the internal queue for log entries to be written.                                           |

Author: The section was removed

2026-03-19 11:53:41.793000,onboarding-server,identityVerification,identityVerification,Switched to PRESENCE_CHECK/VERIFICATION_PENDING; user ID: mockuser_567780649667363606,"{""identityVerificationId"":""189db1e4-7abe-4309-a983-19e6283e3496"",""processId"":""8d0659cb-9e8c-401a-9493-2284f52c5528"",""activationId"":""6c115802-fad3-474a-afe6-b69215740a99"",""userId"":""mockuser_567780649667363606""}"
2026-03-19 11:53:41.740000,onboarding-server,identityVerification,identityVerification,Switched to PRESENCE_CHECK/IN_PROGRESS; user ID: mockuser_567780649667363606,"{""identityVerificationId"":""189db1e4-7abe-4309-a983-19e6283e3496"",""processId"":""8d0659cb-9e8c-401a-9493-2284f52c5528"",""activationId"":""6c115802-fad3-474a-afe6-b69215740a99"",""userId"":""mockuser_567780649667363606""}"
2026-03-19 11:53:41.740000,onboarding-server,presenceCheckProvider,presenceCheckProvider,Presence check started for user: mockuser_567780649667363606,"{""identityVerificationId"":""189db1e4-7abe-4309-a983-19e6283e3496"",""processId"":""8d0659cb-9e8c-401a-9493-2284f52c5528"",""activationId"":""6c115802-fad3-474a-afe6-b69215740a99"",""userId"":""mockuser_567780649667363606""}"

> 2026-03-19 11:53:41.739000,onboarding-server,presenceCheckProvider,presenceCheckProvider,Presence check initialized for user: mockuser_567780649667363606,"{""identityVerificationId"":""189db1e4-7abe-4309-a983-19e6283e3496"",""processId"":""8d0659cb-9e8c-401a-9493-2284f52c5528"",""activationId"":""6c115802-fad3-474a-afe6-b69215740a99"",""userId"":""mockuser_567780649667363606""}"

Author: Fixed, now all columns from the table are included

MESSAGE: `Client evaluated for user: {userId}`

#### Client evaluation accepted

Client evaluation was accepted.

MESSAGE: `Switched to CLIENT_EVALUATION/ACCEPTED; user ID: {userId}`

#### Client evaluation in progress

The [client evaluation service](./External-Onboarding-Services.md#client-evaluation-service) returned `WAIT` — the result is not yet available and will be delivered asynchronously.

MESSAGE: `Switched to CLIENT_EVALUATION/IN_PROGRESS; user ID: {userId}`

#### Client evaluation rejected

The [client evaluation service](./External-Onboarding-Services.md#client-evaluation-service) returned `NOK` — the client was rejected.

MESSAGE: `Switched to CLIENT_EVALUATION/REJECTED; user ID: {userId}`

#### Client evaluation failed

The [client evaluation service](./External-Onboarding-Services.md#client-evaluation-service) failed due to a technical issue.

MESSAGE: `Switched to CLIENT_EVALUATION/FAILED; user ID: {userId}`

#### Client evaluation rejected

Client evaluation was rejected by the [Client evaluation service](./External-Onboarding-Services.md#client-evaluation-service).

> MESSAGE: `Document rejected because of client evaluation for user: {userId}`

Author: Aggregated into single section
